README.md
Allihoopa Desktop SDK
=====================

_Technology preview_

The Allihoopa Desktop SDK provides an interface to the [Allihoopa][] music collaboration service for apps running on MacOS and Windows.

The SDK is a thin `C` interface provided in source form, and has no external dependencies except for what is provided by the C standard library.

The SDK connects to a locally installed Allihoopa App that provides all user interaction and service communication.
This effectively decouples the host app from that complexity, and lets the host app and Allihoopa App have separate update cycles.

## Installation

The API is exposed in the `allihoopa.h` header, and the implementation is in `allihoopa.c`. Simply copy these two into your project and make them compile and link as usual.

>Setting the DEBUG define to a non-zero value will enable some level of tracing to stderr.

## API concepts

To keep the API free of dependencies, and support a wide range of implementation environments, we have set on a couple of simple concepts.

*   **All API calls return an error code**

    Calls will either succeed or fail with an error code.
    A successful call means that a request was successfully sent to the Allihoopa App, and that a response is to be expected.
    Failed calls will not generate responses.

*   **API calls are asynchronous**

    Calls to the SDK will respond quickly, even if the actual request takes a while to execute.
    If the caller is interested in the response, it can be fetched using a separate call.

*   **Complex data is passed as UTF-8 - encoded JSON.**

    This is to avoid having to allocate and manage complex data structures in plain `C`.
    Complex data is always owned by the caller; if you allocated the memory, you free it. *This implies that data returned in callbacks is only valid during that callback, and should be copied if needed.*

*   **Blob files**

    Big binary data, such as audio and image files, are passed as URLs in the JSON data structures. The SDK accepts `http:`, `https:`, `data:` and `file:` protocols. Using the `file:` protocol refers to a client specified temporary directory, and is the recommended method. See the `AHinit` call.

*   **Requests can be tagged with unique IDs**

    API calls accept an integer request ID that is used to tag the request. This is used to match responses to requests, and can be used to free up client side resources associated with a request. The ID must be non-zero, but is not otherwise interpreted by the SDK. *Passing the same non-zero ID to all requests is OK.*

*   **Thread safety**

    The SDK is not thread safe, meaning that all API calls must be made from the same thread. Blocking IO requests between the library and the Allihoopa App will time out after 5 seconds and cause API calls to return `AHErrorCommsFailure`.

## Implementation flow

First, initialize the library by calling `AHinit` with your Allihoopa application ID and API key.

>Contact Allihoopa at developer@allihoopa.com to register your app and get your API key.

You can also provide the root directory for `file:` protocol URLs in the `AHinit` call by setting the `tmpDir` parameter. If not set, this defaults to the system temp directory.

If initialization succeeds, the Allihoopa App window appears, waiting for requests. If the initialization fails with error code `AHErrorAppNotFound`, the Allihoopa App is not installed, or the installation is somehow broken. In that case, direct the user to the `AHSDKHelpURL` url for guidance.

### Dropping

To initiate a minimal drop request, pass a JSON object like this to `AHdrop`:

```json
{
    "stems": {
        "mixStemURL": "file:///my_little_song.wav"
    },
    "presentation": {
        "title": "Singa singa song :)"
    },
    "musicalMetadata": {
        "lengthMicroseconds": 5217392
    }
}
```

If the call succeeds, it will return immediately, and the SDK will start uploading the audio and presenting a dialog for the user.

>NOTE: This is just a minimal example. Please refer to the [pre-release checklist](https://gist.github.com/ReMarkus/ec375c31277cc46cfcc026e69f67c01a) to verify that you’ve integrated our SDK correctly.


### Polling for results

To check the results of previous requests, or to free up temporary resources, call `AHpollCompletedRequests` with a callback function that will be called once for each completed request.

The callback will be called with a JSON object in the following format:

```json
{
    "requestID": 1234,
    "data": {
        ...
    }
}
```

### Queueing up requests and closing the App

Although asynchronous, the SDK will only handle one request at a time. It is OK to initiate a new request before previous requests are handled. The requests will be put in a queue and handled one at a time in order.

>The request and response queues are limited to 100 items each. If the request queue is full, the next request will block. If the reponse queue gets filled up, the oldest response is overwritten.

### More...

This is just a brief overview, see [allihoopa.h](allihoopa.h) for details.

## Allihoopa App installation

The Allihoopa App is the required for the SDK to function, however it is not included in the SDK distribution. The information at the `AHSDKHelpURL` provides information about Allihoopa and guides the end user in downloading and installing the Allihoopa App.

>It is important that the error code `AHErrorAppNotFound` is handled correctly by providing the end user with this information. Simply opening the URL in the system web browser is enough.

[Allihoopa]: https://allihoopa.com


CHANGELOG.md
Change log
==========

## 0.1.0 — 2018-03-23

Technology preview


allihoopa.c
/*

Allihoopa Cross-platform SDK
Copyright 2018 Allihoopa AB

*/


/*

Client code - launches Allihoopa app and communicates using pipes.

Simple IPC used in pipes is based on fixed length messages to
simplify parsing code while being somewhat extensible.

Header:
    4 byte request code ('drop', et.c.)
    2 byte request ID (caller specific, non-zero identifier for the request)
    2 byte unsigned little endian body length (max 65535 bytes)

Body:
    <length> bytes json encoded object

*/

#include "allihoopa.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#if DEBUG
#define TRACE(msg) fprintf(stderr, "%s\n", msg)
#define TRACEF(...) fprintf(stderr, __VA_ARGS__)
#else
#define TRACE(msg)
#define TRACEF(...)
#endif

// Platform specifics with different implementations below
static int readFromApp(char* data, size_t length);
static int writeToApp(const char* data, size_t length);

// Local cross platform glue
static int callApp(
    short int requestID,
    const char* command,
    const char* data, size_t dataLength,
    const char** oBody, size_t* oBodyLength
);

// Exported functions

int AHsetup(const char* setupData, short unsigned int setupDataLength) {
    if (setupData == NULL || setupDataLength == 0) {
        return AHErrorInvalidRequest;
    }
    return callApp(0, "init", setupData, setupDataLength, 0, 0);
}

int AHdrop(const char* dropData, short unsigned int dropDataLength, short int requestID) {
    if (dropData == NULL || dropDataLength == 0 || requestID == 0) {
        return AHErrorInvalidRequest;
    }
    return callApp(requestID, "drop", dropData, dropDataLength, 0, 0);
}

int AHclose() {
    // Since we are closing down the app, ignore but return failed quit requests
    int result = callApp(0, "quit", 0, 0, 0, 0);
    return result;
}

int AHpollCompletedRequests(AHCompletionHandler handler) {
    int moreResults = 0;

    do {
        const char* body = 0;
        size_t bodyLength = 0;

        int pollResult = callApp(0, "poll", 0, 0, &body, &bodyLength);
        if(pollResult == 0) {
            if(bodyLength > AHMaxRequestBody) {
                return AHErrorCommsFailure;
            }
            if(bodyLength != 0) {
                handler(body, bodyLength);
                moreResults = 1;
            }
            else {
                moreResults = 0;
            }
        }
        else {
            return pollResult;
        }
    } while(moreResults);

    return 0;
}

const char* AHerrorCodeToMessage(int errorCode) {
    switch(errorCode) {
        case AHErrorCommsFailure:
            return "App communication failure";
        case AHErrorInvalidRequest:
            return "Invalid request";
        case AHErrorLaunchFailure:
            return "App launch failure";
        case AHErrorOutOfMemory:
            return "Out of memory";
        case AHErrorUnknownError:
        default:
            return "Unknown error";
    }
}

static int callApp(
    short int requestID,
    const char command[4],
    const char* data, size_t dataLength,
    const char** oBody, size_t* oBodyLength)
{
    if(command == 0){
        return AHErrorInvalidRequest;
    }

    if(data == 0 && dataLength != 0){
        return AHErrorInvalidRequest;
    }

    if(data != 0 && dataLength == 0){
        return AHErrorInvalidRequest;
    }

    if(dataLength > AHMaxRequestBody){
        return AHErrorInvalidRequest;
    }

    int result = writeToApp(command, 4);
    if (result) {
        return result;
    }
    result = writeToApp((char*) &requestID, 2);
    if (result) {
        return result;
    }
    result = writeToApp((char*) &dataLength, 2);
    if (result) {
        return result;
    }
    result = writeToApp(data, dataLength);
    if (result) {
        return result;
    }

    char reply[4] = {6, 6, 6, 6};
    result = readFromApp(&reply[0], 4);
    if (result) {
        return result;
    }

    short int responseID = 0;
    result = readFromApp((char*) &responseID, 2);
    if (result) {
        return result;
    }

    if (responseID != requestID) {
        TRACE("Request / response mismatch");
        return AHErrorCommsFailure;
    }

    short unsigned int replyBodyLength = 0;
    result = readFromApp((char*) &replyBodyLength, 2);
    if (result) {
        return result;
    }

    if(replyBodyLength > 0){
        char* body = malloc(replyBodyLength);
        if(body == 0){
            return AHErrorOutOfMemory;
        }
        int bodyResult = readFromApp(body, replyBodyLength);

        if (bodyResult != 0){
            free(body);
            return bodyResult;
        }

        // debug print body here ?

        if (oBody != 0) {
            *oBody = body;
            *oBodyLength = replyBodyLength;
        }
        else {
            free(body);
        }
    }

    static union {
        char chars[4];
        int code;
    } AOK = {
        {'o', 'k', 'a', 'y'}
    };

    TRACEF("Reply: %.4s, %X, %X\n", reply, *(int*) reply, AOK.code);
    if (*(int*) reply != AOK.code) {
        return AHRequestFailed;
    }

    // All OK!
    return 0;
}


/// Platform specific implementations

#ifdef _WIN32
#include <windows.h>

static HANDLE appProcessHandle = 0;
static HANDLE appInputWriteHandle = 0;
static HANDLE appInputReadHandle = 0;
static HANDLE appOutputWriteHandle = 0;
static HANDLE appOutputReadHandle = 0;

/*
 * Creates a pipe with overlapped read, since this is not possible
 * with the regular CreatePipe call.
 * This makes it possible to add timeouts to pipe read, avoiding
 * inter process deadlocks.
 */
static int createPipeWithOverlappedRead(
    HANDLE* oReadHandle,
    HANDLE* oWriteHandle,
    SECURITY_ATTRIBUTES* securityAttributes
)
{
    static int pipeSerial = 0;
    char pipeName[MAX_PATH];

    sprintf(pipeName, "\\\\.\\Pipe\\Allihoopa.%08x.%08x",
        GetCurrentProcessId(),
        pipeSerial++);
    
    const int numPipes = 1;
    const int bufferSize = 8192;
    const int defaultTimeoutMS = 100;

    *oReadHandle = CreateNamedPipe(
        pipeName,
        PIPE_ACCESS_INBOUND | FILE_FLAG_OVERLAPPED,
        PIPE_TYPE_BYTE | PIPE_WAIT,
        numPipes,
        bufferSize,
        bufferSize,
        defaultTimeoutMS,
        securityAttributes
    );

    if(!*oReadHandle){
        return 0;
    }

    *oWriteHandle = CreateFile(
        pipeName,
        GENERIC_WRITE,
        0, // not shared
        securityAttributes,
        OPEN_EXISTING,
        FILE_ATTRIBUTE_NORMAL,
        NULL // no template file
    );

    if(*oWriteHandle == INVALID_HANDLE_VALUE) {
        CloseHandle(*oReadHandle);
        *oReadHandle = 0;
        *oWriteHandle = 0;
        return 0;
    }

    return 1;
}

static int initAppConnection() {
    TRACE("initAppConnection");
    if (appProcessHandle != 0) {
        DWORD exitCode = 0;
        if (GetExitCodeProcess(appProcessHandle, &exitCode) && exitCode == STILL_ACTIVE) {
            TRACE("initAppConnection: reusing live app");
            return 0;
        }
        
        TerminateProcess(appProcessHandle, 0);
        CloseHandle(appProcessHandle);
        appProcessHandle = 0;
    }

    SECURITY_ATTRIBUTES securityAttributes;
    securityAttributes.nLength = sizeof(SECURITY_ATTRIBUTES);
    securityAttributes.bInheritHandle = TRUE;
    securityAttributes.lpSecurityDescriptor = NULL;

    if (appInputReadHandle != 0) {
        CloseHandle(appInputReadHandle);
        appInputReadHandle = 0;
    }
    if (appInputWriteHandle != 0) {
        CloseHandle(appInputWriteHandle);
        appInputWriteHandle = 0;
    }
    if (appOutputReadHandle != 0) {
        CloseHandle(appOutputReadHandle);
        appOutputReadHandle = 0;
    }
    if (appOutputWriteHandle != 0) {
        CloseHandle(appOutputWriteHandle);
        appOutputWriteHandle = 0;
    }

    if(!CreatePipe(&appInputReadHandle, &appInputWriteHandle, &securityAttributes, 0)){
        TRACE("Failed to create app input pipe");
        return AHErrorLaunchFailure;
    }

    if(!SetHandleInformation(appInputWriteHandle, HANDLE_FLAG_INHERIT, 0)) {
        return AHErrorLaunchFailure;
    }

    if(!createPipeWithOverlappedRead(&appOutputReadHandle, &appOutputWriteHandle, &securityAttributes)){
        TRACE("Failed to create app output pipe");
        return AHErrorLaunchFailure;
    }

    if(!SetHandleInformation(appOutputReadHandle, HANDLE_FLAG_INHERIT, 0)) {
        return AHErrorLaunchFailure;
    }

    PROCESS_INFORMATION processInfo;
    ZeroMemory(&processInfo, sizeof(PROCESS_INFORMATION));

    STARTUPINFO startupInfo;
    ZeroMemory(&startupInfo, sizeof(STARTUPINFO));
    startupInfo.cb = sizeof(STARTUPINFO);
    startupInfo.hStdOutput = appOutputWriteHandle;
    // ??? Cannot get redirection to work, using direct console writes from App instead
    //startupInfo.hStdError = stdErrorHandle;
    startupInfo.hStdInput = appInputReadHandle;
    startupInfo.dwFlags |= STARTF_USESTDHANDLES;
    startupInfo.dwFlags |= STARTF_USESHOWWINDOW;
    startupInfo.wShowWindow = SW_SHOWDEFAULT;

    char* cmdLine = "allihoopa.exe -pipe";

    BOOL result = CreateProcess(
        NULL,       // app name
        cmdLine,    // commandline
        NULL,       // security attributes
        NULL,       // thread security attributes
        TRUE,       // inherit handles
        0,          // creation flags
        NULL,       // use parent environment
        NULL,       // use parent's cwd
        &startupInfo,
        &processInfo
        );
    
    if(!result) {
        return AHErrorLaunchFailure;
    }

    appProcessHandle = processInfo.hProcess;
    CloseHandle(processInfo.hThread);

    return 0;
}

static int writeToApp(const char* data, size_t length) {
    TRACE("writeToApp");
    {
        int result = initAppConnection();
        if (result != 0) {
            return result;
        }
    }
    {
        DWORD bytesWritten = 0;
        BOOL result = WriteFile(appInputWriteHandle, data, length, &bytesWritten, 0);
        if ((bytesWritten != length) || !result){
            return AHErrorCommsFailure;
        } else {
            return 0;
        }
    }
}

static int readFromApp(char* data, size_t length) {
    TRACE("readFromApp");
    {
        int result = initAppConnection();
        if (result != 0) {
            return result;
        }
    }
    {
        int ahResult = 0;
        OVERLAPPED overlapInfo;
        ZeroMemory(&overlapInfo, sizeof(OVERLAPPED));
        // We will wait for this event to get signalled on completion
        overlapInfo.hEvent = CreateEvent(
            NULL,
            TRUE, // manual reset
            FALSE, // initial state, not triggered
            NULL);

        if(overlapInfo.hEvent == NULL) {
            return AHErrorUnknownError;
        }

        BOOL readResult = ReadFile(appOutputReadHandle, data, length, 0, &overlapInfo);
        if (!readResult) {
            if (GetLastError() == ERROR_IO_PENDING) {
                DWORD bytesRead = 0;
                int timeoutMS = 1000 * 5;

                int waitResult = WaitForSingleObject(overlapInfo.hEvent, timeoutMS);
                if (waitResult != WAIT_OBJECT_0) {
                    ahResult = AHErrorCommsFailure;
                }
                else {
                    BOOL overlappedResult = GetOverlappedResult(
                        appOutputReadHandle,
                        &overlapInfo,
                        &bytesRead,
                        FALSE // don't block since we waited above
                    );
                    if (!overlappedResult) {
                        // Check GetLastError() == WAIT_TIMEOUT here ?
                        ahResult = AHErrorCommsFailure;
                    }
                }
            } else {
                ahResult = AHErrorCommsFailure;
            }
        }

        CloseHandle(overlapInfo.hEvent);
        return ahResult;
    }
}

#endif // _WIN32

#ifdef __APPLE__
#include "TargetConditionals.h"
#ifdef TARGET_OS_MAC

#include <stdio.h>
#include <fcntl.h>
#include <poll.h>
#include <unistd.h>
#include <errno.h>

static FILE* appPipe = 0;
static int appPipeFD = 0;

static int initAppConnection() {
    if (appPipe != 0) {
        return 0;
    }

    appPipe = popen("./allihoopa -pipe", "r+");
    if (appPipe == 0){
        return AHErrorLaunchFailure;
    }
    appPipeFD = fileno(appPipe);
    if(fcntl(appPipeFD, F_SETFL, O_NONBLOCK) == -1) {
        // The app will be launched and left running
        return AHErrorLaunchFailure;
    }

    return 0;
}

static int waitForData() {
    struct pollfd pollFor = {
        appPipeFD,
        POLLIN,
        POLLIN
    };

    int timeoutMS = 1000 * 5;
    int pollResult = poll(&pollFor, 1, timeoutMS);

    if (pollResult == 1) {
        // one descriptor ready!
        return 0;
    }
    else if(pollResult == 0){
        TRACE("poll timeout!");
        return AHErrorCommsFailure;
    }
    else {
        TRACE("poll failed!");
        return AHErrorCommsFailure;
    }
}

static int readFromApp(char* data, size_t length) {
    int result = initAppConnection();
    if (result) {
        return result;
    }
    
    size_t totalBytesRead = 0;

    do {
        size_t bytesWanted = length - totalBytesRead;
        size_t readResult = read(appPipeFD, &data[totalBytesRead], bytesWanted);

        if (readResult != bytesWanted) {
            if (readResult == 0) {
                // EOF!?
                pclose(appPipe);
                appPipe = 0;
                appPipeFD = 0;
                return AHErrorCommsFailure;
            }
            else if (readResult == -1) {
                if (errno == EAGAIN) {
                    int waitResult = waitForData();
                    if (waitResult) {
                        return waitResult;
                    }
                }
                else if (errno == EINVAL) {
                    return AHErrorCommsFailure;
                }
                else if (errno == EFAULT) {
                    return AHErrorUnknownError;
                }
                else {
                    return AHErrorCommsFailure;
                }
            }
            else {
                totalBytesRead += readResult;
            }
        } else {
            totalBytesRead += readResult;
        }
    }while(totalBytesRead != length);

    return 0;
}

static int writeToApp(const char* data, size_t length) {
    int result = initAppConnection();
    if (result) {
        return result;
    }

    size_t bytesWritten = write(appPipeFD, data, length);
    if (bytesWritten != length) {
        return AHErrorCommsFailure;
    }
    return 0;
}

#endif // TARGET_OS_MAC
#endif // DARWIN


example/testmain.c
#include "allihoopa.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

const char* slurp(const char* path) {
    FILE* file = fopen(path, "r");
    fseek(file, 0, SEEK_END);
    size_t size = ftell(file);
    fseek(file, 0, SEEK_SET);

    char* buffer = malloc(size + 1);
    fread(buffer, size, 1, file);
    buffer[size] = 0;
    return buffer;
}

void waitForEnter() {
    printf("Press enter...");
    getchar();
}

int main() {
    {
        const char* setupRequest = slurp("setup.json");
        assert(setupRequest != NULL);
        int result = AHsetup(setupRequest, strlen(setupRequest));
        if (result) {
            printf("AHsetup returned %d: \"%s\"\n", result, AHerrorCodeToMessage(result));
            return result;
        }
    }

    waitForEnter();

    {
        const char* minimalDropRequest = slurp("minimaldrop.json");
        assert(minimalDropRequest != NULL);
        int result = AHdrop(minimalDropRequest, strlen(minimalDropRequest), 42);
        if (result) {
            printf("AHsetup (minimal) returned %d: \"%s\"\n", result, AHerrorCodeToMessage(result));
            return result;
        }
    }

    waitForEnter();

    if(0){
        const char* fullDropRequest = slurp("fulldrop.json");
        assert(fullDropRequest != NULL);
        int result = AHdrop(fullDropRequest, strlen(fullDropRequest), 43);
        if (result) {
            printf("AHsetup (full) returned %d: \"%s\"\n", result, AHerrorCodeToMessage(result));
            return result;
        }
    }

    if(0){
        int result = AHclose();
        if(result) {
            printf("AHclose returned %d: \"%s\"\n", result, AHerrorCodeToMessage(result));
            return result;        
        }
    }

    return 0;
}


allihoopa.h
/*

Allihoopa Desktop SDK
Copyright 2018 Allihoopa AB

*/

// All API methods accept a JSON formatted request and 
// return a non-zero error code on failure.

#ifndef ALLIHOOPA_H
#define ALLIHOOPA_H
#ifdef __cplusplus
extern "C" {
#endif

/*
    Prepares for Allihoopa communication.
    Must be called before any other API function.

    Setup data format:

    {
        "appID": "<appid>",
        "appKey": "<appkey>",
        "tmpDir": "[optional tempDir where file: - urls are referenced, defaults to system tempdir]"
    }
*/
int AHsetup(const char* setupData, short unsigned int setupDataLength);

/*
    Initiates a drop request

    dropData - utf-8 encoded json object
    requestID - client specific identifier for this request, must not be zero

    returns zero on success, non-zero error code on failure
*/
int AHdrop(const char* dropData, short unsigned int dropDataLength, short int requestID);

/*
    Closes the Allihoopa app.
    Further requests will open a new instance of the app.

    returns zero on success, non-zero error code on failure
*/
int AHclose();

/*
    Callback for receiving results from AHpollCompletedRequests.

    Each response will be json encoded object in the following format:

    {
        "requestID": 1234,
        "data": {
            ...
        }
    }

    responseData: utf-8 encoded json object. Only valid during the duration of the callback.
    responseLength: Length of the responseData in bytes.
*/
typedef void (*AHCompletionHandler)(const char* responseData, unsigned short int responseLength);

/*
    Polls for completed requests.
    Will call the specified handler for each completed request.
*/
int AHpollCompletedRequests(AHCompletionHandler handler);

/*
    Converts from an AHErrors error code to a printable
    string, for logging and debugging.
*/
const char* AHerrorCodeToMessage(int errorCode);

enum AHErrors {
    AHErrorAppNotFound = 451,
    AHErrorLaunchFailure,
    AHErrorCommsFailure,
    AHErrorInvalidRequest,
    AHErrorOutOfMemory,
    AHRequestFailed,
    AHErrorUnknownError,
};

#define AHMaxRequestBody 65535
#define AHSDKHelpURL "https://allihoopa.com/partnerapphelp"

/*

Drop request data, minimal:

{
    "stems": {
        "mixStem": "file:///url/to/audio/data"
    },
    "presentation": {
        "title": "Piece Title"
    },
    "musicalMetadata": {
        "lengthMicroseconds": 123456
    }
}

Drop request data, full format:

{
    "stems": {
        "mixStemURL": "file:///url/to/audio/data"
    },
    "presentation": {
        "title": "Piece Title",
        "previewURL": "file:///optional/url/to/preview/data",
        "coverImageURL": "file:///optional/url/to/cover/image"
    },
    "attribution": {
        "basedOnPieces": [
            "pieceID_A", "pieceID_B"
        ]
    },
    "musicalMetadata": {
        "lengthMicroseconds": 123456,
        "tempo": {
            "fixed": 120.0
        },
        "loop": {
            "startMicroSeconds": 87,
            "endMicroSeconds": 19239
        },
        "timeSignature": {
            "fixed": {
                "upper": 2,
                "lower": 4
            }
        },
        "tonality": {
            "mode": "TONAL",
            "scale": [
                true, false,  true, false,  true,  true, false,  true, false,  true, false,  true
            ],
            "root": 0
        }
    },
    "attachments": [
        {
            "mimeType": "application/x-weird",
            "dataURL": "file:///url/to/attachment/data"
        }
    ]
}

*/

#ifdef __cplusplus
}
#endif

#endif // ALLIHOOPA_H
